doc/tools/docgram/README.md

```diff
@@ -0,0 +1,194 @@
+qualified names such as `Prim.ident`.  A few symbols are defined multiple times,
+such as `scope` and `orient`.
+
+What the tool does
+------------------
+1. The tool reads all the mlg files and generates `fullGrammar`, which includes
+all the grammar without the actions for each production or the OCaml code.  This
+file is provided as a convenience to make it easier to examine the (mostly)
+unprocessed grammar of the mlg files with less clutter.  Nonterminals that use
+levels ("5" RIGHTA below) are modified, for example:
+
+```
+  tactic_expr:
+    [ "5" RIGHTA
+      [ te = binder_tactic -> { te } ]
+```
+
+becomes
+
+```
+tactic_expr5: [
+| binder_tactic
+| tactic_expr4
+]
+```
+
+2. The tool applies grammar editing operations specified by `common.edit_mlg` to
+generate `editedGrammar`.
+
+3. The tool uses `orderedGrammar` to give the desired order for the nonterminals
+and productions in the documented grammar.  `orderedGrammar` has
+the same format as `editedGrammar`.  The tool verifies that `orderedGrammar` is
+a permutation of `editedGrammar`.  Developers are responsible for editing
+`orderedGrammar`.
+
+4. The tool applies further edits to the grammar specified by `prodlist.edit_mlg`,
+then it updates the productionlists in the .mlg files as specified by comments in the form
+`.. insertgram <first nt> <last nt>`.  The edits are primarily to expand
+mlg constructs such as `LIST1` and `OPT` into separate productions.  The tool
+generates `docGrammar`, which has the entire grammar in the form of productionlists.
+
+5. Using the grammar produced in step 3, the tool applies edits specified by
+`prodn.edit_mlg` and generates `prodnGrammar`, representing each production as
+a Sphinx prodn display.  Differently-edited grammars are used because prodn
+can naturally represent `LIST1 x SEP ','` whereas that is awkward for productionlists.
+
+6. TBA: Update the `tacn`, `cmd`, `tacv` and `cmdv` constructs in the .rst files.
+We still need to find a reasonable (i.e. not too tedious) way to associate each of
+these constructs with the production it describes.
+
+How to use the tool
+-------------------
+For now, run `make doc_gram` to update the .rsts and other files.  The plan
+is to eventually do this automatically when a developer does a make or dune build.
+
+The executable takes a list of .mlg and .rst files as arguments (in any order).
+Specifying the `-verify` command line argument avoids updating any of the files,
+but verifies that the current files are consistent.  This setting is meant for
+use in CI; it will be up to each developer include the changes to `*Grammar` and
+the .rst files in their PRs when they've changed the grammar.
+
+Other command line arguments:
+`-verify` described above
+`-fullGrammar` limits processing to updating/verifying only the fullGrammar file.
+`-skip-tacs` supresses warning messages for tactics
+`-skip-cmds` supresses warning messages for commands
```

Comment:
This documentation is confusing. You should distinguish the command-line arguments (when using the tool standalone) and the make targets that developers should know about. In particular, the normal build target should run the tool and regenerate the `fullGrammar` file so that it is easy for developers to commit it as part of their PR without having to think about it. But there should also be a verify target (for this `fullGrammar` file only) that can be run in CI. Finally, there should be another verify target that also verifies the `orderedGrammar` part and the insertion in the manual, but that should be run in a separate allow failure CI job.

Reply:
I created a new section "Command line arguments" to make this clearer.

We can certainly add more make targets but I think that comes a bit later.  I'll elaborate shortly.